pkg/solana/client/client.go

```diff
@@ -0,0 +1,147 @@
+)
+
+//go:generate mockery --name ReaderWriter --output ./mocks/
+type ReaderWriter interface {
+	Writer
+	Reader
+}
+
+type Reader interface {
+	AccountReader
+	Balance(addr solana.PublicKey) (uint64, error)
+	SlotHeight() (uint64, error)
+	LatestBlockhash() (*rpc.GetLatestBlockhashResult, error)
+	ChainID() (string, error)
+	GetFeeForMessage(msg string) (uint64, error)
+}
+
+// AccountReader is an interface that allows users to pass either the solana rpc client or the relay client
+type AccountReader interface {
+	GetAccountInfoWithOpts(ctx context.Context, addr solana.PublicKey, opts *rpc.GetAccountInfoOpts) (*rpc.GetAccountInfoResult, error)
+}
+
+type Writer interface {
+	SendTx(ctx context.Context, tx *solana.Transaction) (solana.Signature, error)
+}
+
+var _ ReaderWriter = (*Client)(nil)
+
+type Client struct {
+	rpc             *rpc.Client
+	skipPreflight   bool // to enable or disable preflight checks
+	commitment      rpc.CommitmentType
+	txTimeout       time.Duration
+	contextDuration time.Duration
+	log             logger.Logger
+
+	// provides a duplicate function call suppression mechanism
+	requestGroup *singleflight.Group
+}
+
+func NewClient(endpoint string, cfg config.Config, requestTimeout time.Duration, log logger.Logger) (*Client, error) {
+	return &Client{
+		rpc:             rpc.New(endpoint),
+		skipPreflight:   cfg.SkipPreflight(),
+		commitment:      cfg.Commitment(),
+		txTimeout:       cfg.TxTimeout(),
+		contextDuration: requestTimeout,
+		log:             log,
+		requestGroup:    &singleflight.Group{},
+	}, nil
+}
+
+func (c *Client) Balance(addr solana.PublicKey) (uint64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.contextDuration)
+	defer cancel()
+
+	v, err, _ := c.requestGroup.Do("GetBalance", func() (interface{}, error) {
+		return c.rpc.GetBalance(ctx, addr, c.commitment)
+	})
+	if err != nil {
+		return 0, err
+	}
+	res := v.(*rpc.GetBalanceResult)
+	return res.Value, err
+}
+
+func (c *Client) SlotHeight() (uint64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.contextDuration)
+	defer cancel()
+	v, err, _ := c.requestGroup.Do("GetSlotHeight", func() (interface{}, error) {
+		return c.rpc.GetSlot(ctx, rpc.CommitmentProcessed) // get the latest slot height
+	})
+	if err != nil {
+		return 0, err
+	}
+	return v.(uint64), nil
+}
+
+func (c *Client) GetAccountInfoWithOpts(ctx context.Context, addr solana.PublicKey, opts *rpc.GetAccountInfoOpts) (*rpc.GetAccountInfoResult, error) {
+	ctx, cancel := context.WithTimeout(ctx, c.contextDuration)
+	defer cancel()
+	opts.Commitment = c.commitment // overrides passed in value - use defined client commitment type
+	return c.rpc.GetAccountInfoWithOpts(ctx, addr, opts)
+}
+
+func (c *Client) LatestBlockhash() (*rpc.GetLatestBlockhashResult, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.contextDuration)
+	defer cancel()
+	return c.rpc.GetLatestBlockhash(ctx, c.commitment)
```

Comment:
Why don't we also use the `requestGroup` here to wrap this call?

Reply:
added: https://github.com/smartcontractkit/chainlink-solana/pull/250/commits/b0b22cdca27331bc30a484ff80a1f706c46ae61e